sm/plugins/tedge_apt_plugin/src/main.rs

```diff
@@ -0,0 +1,131 @@
+use std::process::{Command, Stdio};
+use structopt::StructOpt;
+
+#[derive(StructOpt)]
+struct AptCli {
+    #[structopt(subcommand)]
+    operation: PluginOp,
+}
+
+#[derive(StructOpt)]
+pub enum PluginOp {
+    /// List all the installed modules
+    List {},
+
+    /// Install a module
+    Install {
+        module: String,
+        version: Option<String>,
+    },
+
+    /// Uninstall a module
+    Remove {
+        module: String,
+        version: Option<String>,
+    },
+
+    /// Prepare a sequences of install/remove commands
+    Prepare,
+
+    /// Finalize a sequences of install/remove commands
+    Finalize,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum InternalError {
+    #[error("Fail to run `{cmd}`: {from}")]
+    ExecError { cmd: String, from: std::io::Error },
+}
+
+impl InternalError {
+    pub fn exec_error(cmd: impl Into<String>, from: std::io::Error) -> InternalError {
+        InternalError::ExecError {
+            cmd: cmd.into(),
+            from,
+        }
+    }
+}
+
+fn run(operation: PluginOp) -> Result<std::process::ExitStatus, InternalError> {
+    let status = match operation {
+        PluginOp::List {} => {
+            let apt = Command::new("apt")
+                .args(vec!["--installed", "list"])
+                .stdout(Stdio::piped()) // To pipe apt.stdout into grep.stdin
+                .spawn()
+                .map_err(|err| InternalError::exec_error("apt", err))?;
+
+            let grep = Command::new("grep")
+                .args(vec!["-v", "automatic"])
+                .stdin(apt.stdout.unwrap()) // Cannot panics: apt.stdout has been set
+                .stdout(Stdio::piped()) // To pipe grep.stdout into awk.stdin
+                .spawn()
+                .map_err(|err| InternalError::exec_error("grep", err))?;
+
+            let status = Command::new("awk")
+                .arg(r#"{print "{\"name\":\""$1"\",\"version\":\""$2"\"}"}"#)
+                .stdin(grep.stdout.unwrap()) // Cannot panics: grep.stdout has been set
+                .status()
+                .map_err(|err| InternalError::exec_error("awk", err))?;
+
+            status
+        }
+
+        PluginOp::Install { module, version } => {
+            if let Some(version) = version {
+                run_cmd(
+                    "apt-get",
+                    &format!("install --quiet --yes {}={}", module, version),
+                )?
+            } else {
+                run_cmd("apt-get", &format!("install --quiet --yes {}", module))?
+            }
+        }
+
+        PluginOp::Remove {
+            module,
+            version: _unused,
```

Comment:
We're ignoring the version because `apt` doesn't support the removal of a specific package?

Reply:
Yes.

Reply:
`apt list --manual-installed` and `apt list -installed | grep -v automatic` differ. For instance, `openssl` is listed by the latter but not the former.

```
openssl/focal-security,now 1.1.1f-1ubuntu2.3 amd64 [installed,upgradable to: 1.1.1f-1ubuntu2.4]
```

Reply:
Hmm... Weird. The `--manual-installed` list seems to be more closer to what I've really "manually" installed on my system and is much smaller as well compared to the grepped install list. On my Ubuntu system, the size difference is 69 vs 120. But, I couldn't find any official documentation on the `--manual-installed` option to understand the differences between both, probably because the while `apt list` command is still a WIP even in Debian Buster as per the official man pages [here](https://manpages.debian.org/buster/apt/apt.8.en.html)

So, if a smaller list is our priiority, `--manual-installed` might be a better choice. @rina23q Something that you might wanna consider for your software list shortening work.